### ISSUE TITLE
Upgrade to 1.4.17291.5 on VS 2017 15.4 installs files to wrong directory

### ISSUE THREAD

**Issue by NealGarber** [Microsoft/WindowsTemplateStudio#1323]
## For Bugs:

After upgrading, an error dialog is displayed when opening or closing a solution: "'RelayCommandPackage' package did not load correctly."  I looked in ActivityLog.xml and found the issue is caused by a FileNotFound error looking for C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\Extensions\ipv2fjze.5qr\Microsoft.Templates.2017.dll.  I looked in the ipv2fjze.5qr directory and the only entry is a directory named "~PC".  All other files were deleted by the upgrade. The upgrade appears to delete the files from directory ipv2jfze.5qr and then install the new version to directory inu2pill.sxo but the private registry is not updated to reflect the new location.  So, the files can't be found.  I haven't tried it yet, but I believe a workaround is to copy all of the files/subdirectories from inu2pill.sxo to ipv2fjze.5qr.

Delete of files shown in this attachment:
[dd_setup_20171019091706_001_WindowsTemplateStudio2017.ProRelease.60cee341-5e8e-4db5-a1c4-e3f618500d25.log](https://github.com/Microsoft/WindowsTemplateStudio/files/1400129/dd_setup_20171019091706_001_WindowsTemplateStudio2017.ProRelease.60cee341-5e8e-4db5-a1c4-e3f618500d25.log)

Install of files shown in this attachment:
[dd_setup_20171019091706_003_WindowsTemplateStudio2017.ProRelease.60cee341-5e8e-4db5-a1c4-e3f618500d25.log](https://github.com/Microsoft/WindowsTemplateStudio/files/1400139/dd_setup_20171019091706_003_WindowsTemplateStudio2017.ProRelease.60cee341-5e8e-4db5-a1c4-e3f618500d25.log)

### Repro steps

Upgrade to latest version of the extension.

### Expected Behavior

Install the new version of the extension's files in the same directory or update the private registry so the files can be successfully loaded.

### Actual Behavior

Files were deleted and then reinstalled to a new directory without updating the registry.

## System

VS Version: VS2017 15.4
WTS Wizard Version: N/A
WTS Template Version: N/A
Windows Build: Windows 7 SP1

**Comment by ralarcon** [comment]
Visual studio upgrades the extensions by downloading and installing those in different directories. That's is the regular behaviour. Looks that the upgrade had the assembly locked and could not delete it. 

The question is: does the extension wroked after having restarted Visual Studio? If not, it looks like an update issue and I would recommed to remove the extension and try to install it again.

**Comment by NealGarber** [comment]
Thanks for your suggestion.  I did try an uninstall/reinstall of WTS but he error persisted.  After uninstalling, the error still occurred after opening a project (which tells me it didn't cleanup the private registry).  I'm not confident this is an issue with the extension - I think it's an issue with vsixinstaller.  I reported this in the developer community and someone from MS just responded this morning.  

Also, I've found that uninstalling/reinstalling other extensions has historically placed them in the same Extensions directory.  Perhaps this is because the uninstall doesn't remove the entry from the private registry (which is the case with this issue as well).  In the past, I've compared the directory name on two different PCs that I have and found they are the same for a given extension.  This causes me to believe the "random hash" is really not random, but I don't know the algorithm used.  I can't find documentation on how vsixinstaller derives the "random hash" for the directory name.  Do you know if this is documented?  For grins, can you look on your PC and tell me the directory name where WTS is installed?  I'd be interested to see if it matches one of the names I mentioned (ipv2jfze.5qr or inu2pill.sxo) or if it's completely different.  Thanks again for your assistance..

**Comment by ralarcon** [comment]
This is my folder name: XSS5OCHB.A1F... In different installations I always have seen different folders and after updates I've seen the folder changed...

**Comment by NealGarber** [comment]
Thanks Raúl.  Then it would seem the problem is that vsixinstaller did not completely uninstall the extension during the upgrade and the error was not handled properly.  This doesn't seem like an issue with WTS so I should probably close this issue.  Do you agree?

**Comment by NealGarber** [comment]
More info..  According to http://www.visualstudioextensibility.com/category/vs-15/ 

_"This VS SDK generates two additional files (catalog.json and manifest.json) inside the output .vsix file (which being a .zip file really, you can rename and unzip to inspect) that were not present in the .vsix generated by the VS 2015 SDK"_

Both of these .json files in my vsix have the extensionDir attribute with a value that ends with "inu2pill.sxo".  If a common vsix is downloaded by all, I'm not sure why your installation is different.  Can you explain what I'm missing?

**Comment by ralarcon** [comment]
You are right, the directory for the version 1.4.17291.5 is the same, "inu2pill.sxo" (probably I checked other version).

That said, I've verified that everytime we release a new version (in fact, everytime we create the vsix with the SDK), the directory name is changed.

My recommendation is:
1. Uninstall the Windows Template Studio extension. 
2. Delete both "ipv2fjze.5qr" and "inu2pill.sxo" folders from "C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2017\ENTERPRISE\COMMON7\IDE\EXTENSIONS"  . 
3. Launch Visual Studio and download / install again the Visual Studio Extension extension.

Let me know if this helps.

**Comment by NealGarber** [comment]
I'm giving up trying to find the root cause (it's just taking too much time).  I did the following as a workaround.  We'll see if it fails again with the next version.

1. Uninstall WTS
2. Start regedit, load private registry hive on HKEY_USERS
3. Find and replace all occurrences of ipv2fjze.5qr with uni2pill.sxo
4. Unload private registry hive
5. Reinstall WTS
6. Voila!  The error is gone.  But, I fear I'm doomed to see it again.

I'm not seeing the directory change for other extension (not that I did an exhaustive check) upon upgrade so I'm not sure if this behavior is typical or not.  I can't explain why vsixinstaller doesn't uninstall the RelayCommandPackage when WTS is uninstalled during the upgrade.  I also can't explain why vsixinstaller doesn't reinstall RelayCommandPackage so the CodeBase values can be updated with the new installation directory.  But, at this point, since I wasn't getting any help from MS in the Developer Community, I feel like I'm spinning my wheels and not getting anywhere.  If it fails again with the next upgrade, I'll report the problem again in the Dev. Community.

Thanks for taking the time to try and help me.
